hyperstart/api/grpc/utils.go

```diff
@@ -0,0 +1,140 @@
+package grpc
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func copyValue(to, from reflect.Value) error {
+	toKind := to.Kind()
+	fromKind := from.Kind()
+
+	if !from.IsValid() {
+		return nil
+	}
+
+	if toKind == reflect.Ptr {
+		// If the destination is a pointer, we need to allocate a new one.
+		to.Set(reflect.New(to.Type().Elem()))
+		if fromKind == reflect.Ptr {
+			return copyValue(to.Elem(), from.Elem())
+		} else {
+			return copyValue(to.Elem(), from)
+		}
+	} else {
+		// Here the destination is not a pointer.
+		// Let's check what's the origin.
+		if fromKind == reflect.Ptr {
+			return copyValue(to, from.Elem())
+		}
+
+		switch toKind {
+		case reflect.Struct:
+			return copyStructValue(to, from)
+		case reflect.Slice:
+			return copySliceValue(to, from)
+		case reflect.Map:
+			return copyMapValue(to, from)
+		default:
+			// We now are copying non pointers scalar.
+			// This is the leaf of the recursion.
+			if from.Type() != to.Type() && from.Type().ConvertibleTo(to.Type()) {
+				to.Set(from.Convert(to.Type()))
+				return nil
+			} else {
+				to.Set(from)
+				return nil
+			}
+
+		}
+	}
+}
+
+func copyMapValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Map && from.Kind() != reflect.Map {
+		return fmt.Errorf("Can only copy maps into maps")
+	}
+
+	mapLen := from.Len()
+	to.Set(reflect.MakeMap(to.Type()))
+
+	for j := 0; j < mapLen; j++ {
+		if err := copyValue(to.Index(j), from.Index(j)); err != nil {
+			return err
```

Comment:
I think this code will panic. Value.Index(int) cannot be used on maps.  There are MapIndex and SetMapIndex functions that can be used instead.